File: src/j/implementation/microSet/InnerSet_1.java
package j.implementation.microSet;

import java.util.Iterator;

public class InnerSet_1 <T> extends AbstractInnerSet<T>
{
        protected T element_1;
        
    public InnerSet_1(T p_element_1)
    {
                element_1 = p_element_1;
            }

    public InnerSet_1(InnerSet_1<T> i)
    {
            element_1 = i.element_1;
        }

    public InnerSet_1<T> copy() {
        return this;
    }

    public InnerSet_1(InnerSet_0<T> i, T el)
    {
                    element_1 = el;
    }
    
    @Override
    public InnerSet<T> addElement(T p_element)
    {
        if(element_1.equals(p_element))
            return this;
        else
            return new InnerSet_2<>(this, p_element);    }

    @Override
    public InnerSet<T> addUnChecked(T p_element)
    {
        return new InnerSet_2<>(this, p_element);    }

    @Override
    public InnerSet<T> removeElement(Object p_element)
    {
                if(element_1.equals(p_element))
            return  (InnerSet_0<T>) InnerSet_0.singleton
            ;
                        else
            return this;
    }

    @Override
    public boolean containsElement(Object p_element)
    {
        return element_1.equals(p_element);
    }


    public T getElement(int index)
    {switch(index) {
                case 0:
            return element_1;
                default:
            return null;
            }
    }


    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
        // Reverse the add if the innerSet we add is smaller than this
        if (innerSet.getSize() < this.getSize()) {
            return innerSet.addAllElements(this);
        }
        return innerSet.addElement(element_1);
    }


    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
        return innerSet.removeElement(element_1);
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
        InnerSet<T> i = this;
                if (!innerSet.containsElement(element_1)) i = i.removeElement(element_1);
                return i;
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        return  innerSet.containsElement(element_1);
    }

    public InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAllAndPropagateReverse(this, microSetToPropagate);
    }


    public InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAndPropagate(element_1, microSetToPropagate);
    }

    public InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSetToPropagate) {
        if(element_1.equals(p_element))
            return this;
        else {
            microSetToPropagate.add(p_element);
            return new InnerSet_2<>(this, p_element);        }
    }

    @Override
    public String toString()
    {
        return "{" + element_1 + " }";
    }
    
    @Override
    public int getSize()
    {
        return 1;
    }

    @Override
    public Iterator<T> iterator() {
        return new InnerSetIterator<>(this);
    }

    @Override
    public InnerSet<T> clear(boolean unused) {
        return (InnerSet<T>) InnerSet_0.singleton;
    }
}

File: src/cpp/implementation/save/generate/inner_set_1.cpp
#include "inner_set_1.hpp"

template<typename T, int p_to_up>
inner_set_1<T, p_to_up>::inner_set_1()
{

}

template<typename T, int p_to_up>
inner_set_1<T, p_to_up>::inner_set_1(T p_element_1)
{
            a_values.element_1 = p_element_1;
    }

template<typename T, int p_to_up>
inner_set_1<T, p_to_up>::inner_set_1(inner_set_0<T, p_to_up>* p_inner_set, T p_element)
{
                    a_values.element_1 = p_element;
    
	delete p_inner_set;
}

template<typename T, int p_to_up>
inner_set_1<T, p_to_up>::~inner_set_1()
{

}

template<typename T, int p_to_up>
inner_set<T, p_to_up>*  inner_set_1<T, p_to_up>::copy()
{
	return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::add_element(T p_element)
{
	if(!contains_element(p_element))
		if(this->a_to_up == 1)
			return new inner_array_set<T, p_to_up>(this, p_element);
		else
			return new inner_set_2<T, p_to_up>(this, p_element);
	else
		return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::add_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->add_element(a_values.element_1);
}

template<typename T, int p_to_up>
bool inner_set_1<T, p_to_up>::contains_element(T p_element)
{
    return  a_values.element_1 == p_element;
}

template<typename T, int p_to_up>
bool inner_set_1<T, p_to_up>::contains_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
   return  p_inner_set->contains_element(a_values.element_1) ;
}

template<typename T, int p_to_up>
T inner_set_1<T, p_to_up>::get_element(int p_index)
{
    return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::remove_element(T p_element)
{
                if(a_values.element_1 == p_element)
                return inner_set_0<T, p_to_up>::EMPTY;
        ;
    	}
                        else
            return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::remove_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->remove_element(a_values.element_1);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::retain_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  inner_set<T, p_to_up>* r_inner_set = inner_set_0<T, p_to_up>::EMPTY;

		if(p_inner_set->contains_element(a_values.element_1))
		 r_inner_set = r_inner_set->add_element(a_values.element_1);
	
	 
    if(this->get_size() != 0)
      delete this;

  return r_inner_set;
}


template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::add_and_propagate(T p_element, micro_set<T, p_to_up>* p_micro_set)
{
	if( a_values.element_1 == p_element )
		return this;
	else
  	{
  		p_micro_set->insert(p_element);

		return new inner_set_2<T, p_to_up>(this, p_element);
	}
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::add_all_and_propagate(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
  return p_inner_set->add_all_and_propagate_reverse(this, p_micro_set);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_1<T, p_to_up>::add_all_and_propagate_reverse(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
	return p_inner_set->add_and_propagate(a_values.element_1, p_micro_set);
}

template<typename T, int p_to_up>
int inner_set_2<T, p_to_up>::get_size()
{
    return 1;
}
